addons/nuxeo-retention/nuxeo-retention-core/src/main/java/org/nuxeo/retention/adapters/Record.java

```diff
@@ -0,0 +1,139 @@
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.IdRef;
+import org.nuxeo.ecm.core.api.versioning.VersioningService;
+import org.nuxeo.ecm.platform.audit.service.NXAuditEventsService;
+import org.nuxeo.ecm.platform.dublincore.listener.DublinCoreListener;
+import org.nuxeo.ecm.platform.ec.notification.NotificationConstants;
+import org.nuxeo.retention.RetentionConstants;
+
+/**
+ * @since 11.1
+ */
+public class Record {
+
+    private static final Logger log = LogManager.getLogger(Record.class);
+
+    protected DocumentModel document;
+
+    public Record(final DocumentModel doc) {
+        document = doc;
+    }
+
+    public DocumentModel getDocument() {
+        return document;
+    }
+
+    public RetentionRule getRule(CoreSession session) {
+        List<String> ruleIds = getRuleIds();
+        if (ruleIds == null) {
+            return null;
+        }
+        for (String ruleId : ruleIds) {
+            IdRef ruleRef = new IdRef(ruleId);
+            if (!session.exists(ruleRef)) {
+                log.trace("Rule {} does not exist", ruleRef);
+                continue;
+            }
+            DocumentModel ruleDoc = session.getDocument(ruleRef);
+            RetentionRule rule = ruleDoc.getAdapter(RetentionRule.class);
+            if (!rule.isEnabled()) {
+                log.debug("Rule {} id disabled", ruleDoc::getPathAsString);
+                continue;
+            }
+            return rule;
+        }
+        log.debug("No active rules found for {}", document::getPathAsString);
+        return null;
+
+    }
+
+    protected List<String> getRuleIds() {
+        Serializable propertyValue = document.getPropertyValue(RetentionConstants.RECORD_RULE_IDS_PROP);
+        if (propertyValue == null) {
+            return Collections.emptyList();
+        }
+        return Arrays.asList((String[]) propertyValue);
+    }
+
+    public Calendar getSavedRetainUntil() {
+        Serializable savedRetainUntil = document.getPropertyValue(RetentionConstants.RETAIN_UNTIL_PROP);
+        if (savedRetainUntil != null) {
+            return (Calendar) savedRetainUntil;
+        }
+        return null;
+    }
+
+    public boolean isRetainUntilInderterminate() {
+        if (!getDocument().isUnderRetentionOrLegalHold()) {
+            return false;
+        }
+        Calendar retainUntil = getDocument().getRetainUntil();
+        return retainUntil != null
+                ? CoreSession.RETAIN_UNTIL_INDETERMINATE.getTimeInMillis() == retainUntil.getTimeInMillis()
+                : false;
+    }
+
+    public boolean isRetentionExpired() {
+        if (!getDocument().isUnderRetentionOrLegalHold()) {
+            return true;
+        }
+        Calendar retainUntil;
+        return (retainUntil = getDocument().getRetainUntil()) == null || !Calendar.getInstance().before(retainUntil);
```

Comment:
i am wondering why we don't use directly the `getDocument().isUnderRetentionOrLegalHold()` because the second check  `(retainUntil = getDocument().getRetainUntil()) == null || !Calendar.getInstance().before(retainUntil);` is already covered by `getDocument().isUnderRetentionOrLegalHold()` or perhaps the need is to be sure that we was under retention and the data now is in the past

Reply:
This checks only retention expiration, whereas `isUnderRetentionOrLegalHold` can also return `true` when there's a legal hold.